runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java

```diff
@@ -1864,6 +1890,22 @@ private void sendWorkerUpdatesToDataflowService(
     }
   }
 
+  public void handleHeartbeatResponses(List<Windmill.ComputationHeartbeatResponse> responses) {
+    for (Windmill.ComputationHeartbeatResponse computationHeartbeatResponse : responses) {
+      Map<Long, List<FailedTokens>> failedWork = new HashMap<>();
+      for (Windmill.HeartbeatResponse heartbeatResponse :
+          computationHeartbeatResponse.getHeartbeatResponsesList()) {
+        failedWork.putIfAbsent(heartbeatResponse.getShardingKey(), new ArrayList<>());
+        failedWork
+            .get(heartbeatResponse.getShardingKey())
```

Comment:
putIfAbsent returns the value so you can avoid this get

Reply:
It looks like it returns null if the key was not already in the Map, but I think I got it to work.